server/routes/api/documents/documents.ts

```diff
@@ -435,6 +436,59 @@ router.post(
   }
 );
 
+router.post(
+  "documents.users",
+  auth(),
+  pagination(),
+  validate(T.DocumentsUsersSchema),
+  async (ctx: APIContext<T.DocumentsUsersReq>) => {
+    const { id, query } = ctx.input.body;
+    const actor = ctx.state.auth.user;
+    const { offset, limit } = ctx.state.pagination;
+    const document = await Document.findByPk(id);
+    authorize(actor, "read", document);
+
+    let users: User[] = [];
+    let total = 0;
+
+    if (document.collectionId) {
+      const collection = await Collection.findByPk(document.collectionId);
+      authorize(actor, "update", collection);
+      const memberIds = await Collection.membershipUserIds(collection.id);
```

Comment:
```suggestion
      const [collection, memberIds] = Promise.all([
        Collection.findByPk(document.collectionId),
        Collection.membershipUserIds(document.collectionId)
      ]);
      authorize(actor, "update", collection);
```

I think it would be worth running these in parallel